### PR TITLE
Support for Monitoring and Tracing

### PR DESCRIPTION
**Pull request by juzuluag**
Issue Number: 109

## Summary
Added support for WebAPI Monitoring and Telemetry

## Changes
- Added code to hook Application Insights with WebApi service
- Added consistent tracing mechanism across both Controllers

## Checklist

- [x] Local Tests Passing?
- [x] CICD and Pipeline Tests Passing?
- [ ] Added any new Tests?
- [x] Documentation Updates Made?
- [ ] Are there any API Changes? If yes, please describe below.
- [x] This is not a breaking change. If it is, please describe it below.

## Are there API Changes?
If yes, what are the expected API Changes? Please link to an API-Comparison workflow with the API Diff.

## Is this a breaking change?
If yes, what workflow does this break? 

## Anything else?
This is integrated with Azure AppInsights based on the current customer scenario.

### CODE REVIEW

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs
Comment (JenMadiedo) [inline src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs:28]: What was the rationale for starting this here? I have vague memories that @bderusha was very particular about where he initialized the Activity source but I'm not familiar with that reasoning

Comment (JenMadiedo) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs:28]: On second thought the important piece may be where it's started, not where it's initialized, so large grain-of-🧂 with this question^^

File: src/CarbonAware.LocationSources/CarbonAware.LocationSources.Azure/src/AzureLocationSource.cs
Comment (JenMadiedo) [inline src/CarbonAware.LocationSources/CarbonAware.LocationSources.Azure/src/AzureLocationSource.cs:62]: thanks for the null catch!

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs
Comment (juzuluag) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs:28]: Hi @JenMadiedo, followed this [doc](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/distributed-tracing-instrumentation-walkthroughs#best-practices-1)

**Review by JenMadiedo — APPROVED** [review]
Changes look good to me but I'm in no way an expert on when/where activities should be started for logging so I'll let Robert/Bill comment on that

File: src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs
Comment (juzuluag) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs:28]: BTW, some of these changes are based on the fact that there was a lot of inconsistencies along the code based related with activities.

Comment (JenMadiedo) [reply src/CarbonAware.DataSources/CarbonAware.DataSources.WattTime/src/WattTimeDataSource.cs:28]: Ahh nice, love best practices docs 🤓 Thanks for bringing up to consistency

**Review by KanishkT123 — CHANGES_REQUESTED** [review]
Had a call with @juzuluag and @pritipath 

Changes requested: We need to have some kind of `switch` or config reading statement so that AppInsights is not the only/default option. This would follow the same design philosophy as the Datasource and Aggregators, allowing for future work to be more extensible. Otherwise, we lock everyone into the choice of Azure, which is contrary to the design decisions already made else where.

**Review by bderusha — COMMENTED** [review]
I agree with @KanishkT123 about aligning this config with the patterns established in the rest of the codebase.

My other comment is a bit more broad.  What's written in this PR is great.  There is one additional thing I was hoping/expecting this PR to address related to this [comment from the MS docs](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/distributed-tracing-instrumentation-walkthroughs#optional-add-additional-activities):

> In a larger and more realistic project, using an Activity in every method would produce extremely verbose traces, so it's not recommended.

I would love to have a documented opinion on how we think about starting new activities.
- Should it be all public methods? Not true for WattTimeClient, which only traces a few private methods.
- Should it be configurable by the user? `using (var activity = config.TraceAggregators ? Activity.StartActivity() : null)`
- Only trace at controllers methods and disk/network methods? To optimize for insight into higher-latency functions.

These answers will help guide us and future contributors on when/where it is appropriate to add traces when writing new code.

**Comment by juzuluag** [comment]
> I agree with @KanishkT123 about aligning this config with the patterns established in the rest of the codebase.
> 
> My other comment is a bit more broad. What's written in this PR is great. There is one additional thing I was hoping/expecting this PR to address related to this [comment from the MS docs](https://docs.microsoft.com/en-us/dotnet/core/diagnostics/distributed-tracing-instrumentation-walkthroughs#optional-add-additional-activities):
> 
> > In a larger and more realistic project, using an Activity in every method would produce extremely verbose traces, so it's not recommended.
> 
> I would love to have a documented opinion on how we think about starting new activities.
> 
> * Should it be all public methods? Not true for WattTimeClient, which only traces a few private methods.
> * Should it be configurable by the user? `using (var activity = config.TraceAggregators ? Activity.StartActivity() : null)`
> * Only trace at controllers methods and disk/network methods? To optimize for insight into higher-latency functions.
> 
> These answers will help guide us and future contributors on when/where it is appropriate to add traces when writing new code.

Great, I was waiting for these type of questions, since currently `ActivitySource` use still very sporadically and it is not consistent. 
Could we do this as part of another story, to do the consensus on how to use it? I'm more towards the later, meaning only use it for I/o bounds and Controller entry points. The rest is not beneficial imo, and it would make another PR small and better scoped.

File: GettingStarted.md
Comment (bderusha) [inline GettingStarted.md]:
```suggestion
Application monitoring and tracing can be configured using the `TELEMETRY_PROVIDER` variable in the application configuration. 
```

File: GettingStarted.md
Comment (bderusha) [inline GettingStarted.md]:
```suggestion
This application is integrated with Application Insights for monitoring purposes. The telemetry collected in the app is pushed to AppInsights and can be tracked for logs, exceptions, traces and more. To connect to your Application Insights instance, configure the `APPLICATIONINSIGHTS_CONNECTION_STRING` variable:
```

File: src/CarbonAware.WebApi/src/Program.cs
Comment (bderusha) [inline src/CarbonAware.WebApi/src/Program.cs]: Consider using a constant for these values

File: src/CarbonAware.WebApi/src/Program.cs
Comment (bderusha) [inline src/CarbonAware.WebApi/src/Program.cs]: Consider making this case-insensitive

File: src/CarbonAware.WebApi/src/Program.cs
Comment (bderusha) [reply src/CarbonAware.WebApi/src/Program.cs]: I wonder if it would be worth using a similar pattern to what we do for [DataSource Registration]( https://github.com/microsoft/carbon-aware-sdk/blob/dev/src/CarbonAware.DataSources/CarbonAware.DataSources.Registration/Configuration/ServiceCollectionExtensions.cs) to contain this logic.

**Comment by bderusha** [comment]
> Great, I was waiting for these type of questions, since currently `ActivitySource` use still very sporadically and it is not consistent. Could we do this as part of another story, to do the consensus on how to use it? I'm more towards the later, meaning only use it for I/o bounds and Controller entry points. The rest is not beneficial imo, and it would make another PR small and better scoped.

@juzuluag I think it probably makes sense to address as part of this story, but I would defer to @esgraham as to how to organize the work.  I do agree that it can/should be a follow-on PR and doesn't block this from merging.

**Review by bderusha — APPROVED** [review]
Looks great!  :shipit:

**Review by KanishkT123 — APPROVED** [review]
It looks great to me! Agree with Bill on needing a new story for Activity traces, but other than that, this is great. 

The Integration Tests have broken, but that's not because of this story.

**Comment by juzuluag** [comment]
> It looks great to me! Agree with Bill on needing a new story for Activity traces, but other than that, this is great.
> 
> The Integration Tests have broken, but that's not because of this story.

@KanishkT123 , good to know about the Integration Test. Seems like w/o fixing it, this PR can't be merge.

**Comment by juzuluag** [comment]
> > It looks great to me! Agree with Bill on needing a new story for Activity traces, but other than that, this is great.
> > The Integration Tests have broken, but that's not because of this story.
> 
> @KanishkT123 , good to know about the Integration Test. Seems like w/o fixing it, this PR can't be merge.

Oh sorry, it can be. Thanks.

**Pull request merged**
